### DISCUSSION TITLE
Feature Request: Better ZodObject generics

### DISCUSSION THREAD

**Discussion by geoffgscott** [colinhacks/zod#2102] [Q&A]
In TS projects I find I am frequently defining schemas based on an existing type using this process:

```ts
const userSchema: ZodSchema<User> = z.object({
    name: z.string(),
    email: z.string().email()
})
```

However the generic `ZodSchema` strips all the `ZodObject` properties like `partial` and `shape` from the type so when writing functions that take a ZodObject we need to typecast from schema to object. 

It would be very helpful to have a `ZodSchemaObject` generic that would map to type with all the properties of `ZodObject`

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const userSchema = z.object({
    name: z.string(),
    email: z.string().email()
}) satisfies z.ZodSchema<User>

userSchema.shape // no compile error
userSchema.partial // no compile error
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by geoffgscott** [comment]
I don't think so because the intent would be to do something like this:

```ts
function foo<T extends object>(schema: ZodObject<T>){
   return schema.shape
}
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
function foo<Schema extends z.AnyZodObject> ( schema: Schema ) {
    return schema.shape
}
```

**Comment by geoffgscott** [comment]
No. AnyZodObject will strip the generic information and restrict the ability to reference the base type that drives the schema. A better use case might be this:

```ts
function keyOrder<T extends object>(schema: ZodObject<T>){

  const numberedKeys: Partial<{[x in keyof T]: number}> = {}

  Object.keys(schema.shape).forEach((key, idx) => {
    numberedKeys[key] = idx
  })

   return numberedKeys as {[x in keyof T]: number}
}
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
function keyOrder<T extends z.ZodRawShape> ( schema: z.ZodObject<T> ) {
    const keys = Object.keys( schema.shape ) as ( keyof T )[]

    const numberedKeys = keys.reduce(
        ( numberedKeys, key, index ) => ( { ...numberedKeys, [ key ]: index } ),
        {} as { [ Key in keyof T ]: number }
    )

    return numberedKeys
}

const foo = keyOrder(
    z.object( {
        a: z.string(),
        b: z.string(),
        c: z.string(),
    } )
)

console.log( foo ) // { a: 0, b: 1, c: 2 }
```

**Reply by geoffgscott** [reply]
No. ZodRawShape would not be representative of the actual type. 

Something like:
```
type User = {
   name: string,
   email: string
}
```

**Reply by JacobWeisenburger** [reply]
So are you basically asking for a better way to make `ZodRawShape` work with a typical type such as `type User = { name: string, email: string }`?

**Reply by geoffgscott** [reply]
Ideally in the ZodObject generic it would take a standard type and not a ZodRawShape (or possibly either and handle the casting from `User` to `ZodRawShape<User>`)

**Reply by geoffgscott** [reply]
Is this maybe the best practice?

```ts
const userSchema: ZodObject<any, any, any, User> = z.object({
  name: z.string()
})
```

And then not the input to the functions can be `ZodObject<any, any, any, T>`

**Reply by geoffgscott** [reply]
I can't say I find the above a very elegant solution. It would be really nice to have something like a `ZodObjectSchema` which extends `ZodSchema` but includes the properties of ZodObject.

I am also noticing that while the above maintains types for basic parsing it drops them when using things like `.partial()`
